common/tasks/sequential_scheduler.go

```diff
@@ -0,0 +1,328 @@
+
+		options      *SequentialSchedulerOptions
+		queues       collection.ConcurrentTxMap
+		queueFactory SequentialTaskQueueFactory[T]
+		queueChan    chan SequentialTaskQueue[T]
+
+		logger log.Logger
+	}
+)
+
+func NewSequentialScheduler[T Task](
+	options *SequentialSchedulerOptions,
+	taskQueueHashFn collection.HashFunc,
+	taskQueueFactory SequentialTaskQueueFactory[T],
+	logger log.Logger,
+) *SequentialScheduler[T] {
+	return &SequentialScheduler[T]{
+		status:       common.DaemonStatusInitialized,
+		shutdownChan: make(chan struct{}),
+		options:      options,
+
+		logger: logger,
+
+		queueFactory: taskQueueFactory,
+		queueChan:    make(chan SequentialTaskQueue[T], options.QueueSize),
+		queues:       collection.NewShardedConcurrentTxMap(1024, taskQueueHashFn),
+	}
+}
+
+func (s *SequentialScheduler[T]) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	s.startWorkers(s.options.WorkerCount())
+
+	s.shutdownWG.Add(1)
+	go s.workerMonitor()
+
+	s.logger.Info("sequential scheduler started")
+}
+
+func (s *SequentialScheduler[T]) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(s.shutdownChan)
+	// must be called after the close of the shutdownChan
+	s.drainTasks()
+
+	go func() {
+		if success := common.AwaitWaitGroup(&s.shutdownWG, time.Minute); !success {
+			s.logger.Warn("sequential scheduler timed out waiting for workers")
+		}
+	}()
+	s.logger.Info("sequential scheduler stopped")
+}
+
+func (s *SequentialScheduler[T]) Submit(task T) {
+	queue := s.queueFactory(task)
+	queue.Add(task)
+
+	_, fnEvaluated, err := s.queues.PutOrDo(
+		queue.ID(),
+		queue,
+		func(key interface{}, value interface{}) error {
+			value.(SequentialTaskQueue[T]).Add(task)
+			return nil
+		},
+	)
+	if err != nil {
+		panic("Error is not expected as the evaluation function returns nil")
+	}
+
+	// if function evaluated, meaning that the task set is
+	// already dispatched
+	if fnEvaluated {
+		if s.isStopped() {
+			s.drainTasks()
+		}
+		return
+	}
+
+	// need to dispatch this task set
+	select {
+	case <-s.shutdownChan:
+	case s.queueChan <- queue:
+		if s.isStopped() {
+			s.drainTasks()
+		}
+	}
+}
+
+func (s *SequentialScheduler[T]) TrySubmit(task T) bool {
+	queue := s.queueFactory(task)
+	queue.Add(task)
+
+	_, fnEvaluated, err := s.queues.PutOrDo(
+		queue.ID(),
+		queue,
+		func(key interface{}, value interface{}) error {
+			value.(SequentialTaskQueue[T]).Add(task)
+			return nil
+		},
+	)
+	if err != nil {
+		panic("Error is not expected as the evaluation function returns nil")
+	}
+	if fnEvaluated {
+		if s.isStopped() {
+			s.drainTasks()
+		}
+		return true
+	}
+
+	select {
+	case s.queueChan <- queue:
+		if s.isStopped() {
+			s.drainTasks()
+		}
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *SequentialScheduler[T]) workerMonitor() {
```

Comment:
the ability to dynamically start / stop worker should be extracted into a new util, but maybe later